### PR TITLE
joystick support for VCMI

### PR DESCRIPTION
**Pull request by kdmcser**
# Development Purpose

In some handheld game device such as Stream Deck or Odin2, touching screen may not be the best way to play VCMI. And someone who like connect his PC to TV, sitting in sofa and use a joystick will make him more comfortable. Unfortunely, I happen to fit both situations. So I have to write those codes.

# Design

libSDL2 support mapping  joysticks to a unified control scheme: game controller. So some game controller event should be listened. I add device events to preprocessing event so that they can be listened during intro video playing. And add axis and button event into current event. button event should also ending video playing.

After main menu is showed, I fetch the event of left axis move, and calculate the pointer move distance by each frame. bottons and triggers is design to trigger some shortcuts and simulate mouse right click.

# Operation Binding

This this my first step, I simply mapping axis and button to these features:

**Left Stick:** Move cursor
**Right Stick:** None
**A:** Left mouse click
**B:** ESC
**X:** Enter
**Y:** Right mouse click
**LB :** Adventure map hero switch, defense
**RB:** Adventure map town switch, wait
**LT :** Access current building, tactical next, creature spell
**RT :** Hero spell
**Back:** End turn, auto combat until battle end
**Start:** Open options menu, open combat options menu
**D-pad up:** Upward movement, view world, unit upgrade, upgrade all Highland Forts, stack all skeletons, recruit maximum, scroll battle info upwards
**D-pad down**: Downward movement, view kingdom overview, recruit minimum, scroll battle info downwards
**D-pad left:** Leftward movement, view scene info
**D-pad right:** Rightward movement, view thieves guild info
**L3 (Left stick press):** Switch surface/underground, toggle hero info during battle
**R3 (Right stick press):** Toggle grid (not implemented), toggle action queue during battle

# Future Work

This commit is only a solve 0 to 1 problem. Next step is making operation binding modable by modder or user.

### CODE REVIEW

**Comment by kdmcser** [comment]
> why have you closed the PR?

compile failed when update today code. I fix it and reopen

**Comment by kambala-decapitator** [comment]
no need to close, just convert the PR to draft

**Comment by kdmcser** [comment]
I only see the Close button，didn't see the convert the PR to draft link before. Now i find it.

**Comment by dydzio0614** [comment]
i wonder if racing wheels will work with this pull request :smile:

**Comment by IvanSavenko** [comment]
Works fine with my DualShock controller. Definitely quite good for first attempt. Some issues:
- cursor is only visible if you have software cursor selected in Launcher. 
- D-pad does not works for hero movement for me. It does other stuff that you've listed, e.g. kingdom overview, but not hero movement
- there is no way to select option in selection dialogs other than cursor - chest pickup, hero levelup, school of magic, etc
- button assignments are somewhat inconvenient, but not sure what to do considering H3 control scheme. For example, on levelup window 'B' / circle has no function at all, you need to press X / square to confirm or move mouse pointer and use 'A' / cross

Have you considered some alternative layouts? For example:
- LB: left mouse button
- RB: right mouse button
- 'A' / cross: confirmation (Return)
- 'B' / circle: cancel (Esc)
- LT: select first option in dialogs
- RT: select second / last option in dialogs

**Comment by kdmcser** [comment]
> cursor is only visible if you have software cursor selected in Launcher.
cursor problem I will solve soon.

> D-pad does not works for hero movement for me. It does other stuff that you've listed, e.g. kingdom overview, but not hero movement
I often mispress the D pad when using left joystick , so I change it features to list before. I will make buttons feature configurable in next version. 

> there is no way to select option in selection dialogs other than cursor - chest pickup, hero levelup, school of magic, etc
Is any shortcut can do this job? or need to develop more?

> button assignments are somewhat inconvenient, but not sure what to do considering H3 control scheme. For example, on levelup window 'B' / circle has no function at all, you need to press X / square to confirm or move mouse pointer and use 'A' / cross

I just set be to EShortcut::GLOBAL_CANCEL, EShortcut::GLOBAL_RETURN and EShortcut::ADVENTURE_EXIT_WORLD_VIEW
the GLOBAL_RETURN should not be set?

**Comment by IvanSavenko** [comment]
> Is any shortcut can do this job? or need to develop more?

I think you can use SELECT_INDEX_1 / SELECT_INDEX_2 shortcuts here.
They will select first / second option in such dialogs. Should be enough for OH3 which does not have windows with more than two options.

> I just set be to EShortcut::GLOBAL_CANCEL, EShortcut::GLOBAL_RETURN and EShortcut::ADVENTURE_EXIT_WORLD_VIEW
the GLOBAL_RETURN should not be set?

I think main problem here is that button assignment is different from how games typically use controllers.
If I play with controller, I expect 'A' / cross to act as "confirm" and 'B' / circle to act as "cancel".

However you use 'A' as left mouse click which is not how most of games use this button.

Using scheme that I have described above, with mouse buttons mapped to LT/RT will make it possible to use A/B buttons for their expected function - confirm and cancel

**Comment by kdmcser** [comment]
Binding A to Enter and B to ESC seems a better idea, but binding LB to left mouse button maybe a little  inconvenient. that means you need to use your left hand to move  and use left hand to click. So bind left mouse button to X and right mouse button to Y maybe better?

**Comment by IvanSavenko** [comment]
> Binding A to Enter and B to ESC seems a better idea, but binding LB to left mouse button maybe a little inconvenient. that means you need to use your left hand to move cursor and use left hand to click. 

What about allowing to use both left & right sticks for mouse movement - whatever is more convenient to a player? Or is using right hand for mouse movement and left hand for mouse click is also inconvenient? 

Sorry, I am left-handed, so some of your common sense might not be common sense to me as far as controls go.

> So bind left mouse button to X and right mouse button to Y maybe better?

Not sure if this is good option, but yeah, probably better than current one.


By the way - make sure to grab latest changes from develop branch. I actually made keyboard shortcuts configurable in json few days ago - check this PR: #3749
You might be able to reuse some of these changes for configurable controller mappings. I used `SDL_GetKeyName` for key -> string mapping, and looks like SDL has `SDL_GameControllerGetStringForButton` for controllers that you can use for same effect

**Comment by kdmcser** [comment]
So make it configurable for different users is a good solution. no matter left-handed or right-handden. I will prepair a config file in config/joystick.json to make users costumize there button binding as #3749 did

**Comment by IvanSavenko** [comment]
Yeah, probably the best option.
You can make it as part of `config/shortcutsConfig.json`, and load it in ShortcutHandler class. Don't see need for a separate config/class

File: client/eventsSDL/InputHandler.cpp [hunk 1]
```diff
@@ -196,6 +206,21 @@ void InputHandler::preprocessEvent(const SDL_Event & ev)
...
+    else if(ev.type == SDL_CONTROLLERDEVICEREMOVED)
+    {
+        gameControllerHandler->handleEventDeviceRemoved(ev.cdevice);
+        return;
+    }
+    else if(ev.type == SDL_CONTROLLERDEVICEREMAPPED)
+    {
+        gameControllerHandler->handleEventDeviceRemapped(ev.cdevice);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (kdmcser) [reply client/eventsSDL/InputHandler.cpp:223]: I first put it there, but if joystick is connected while intro video played. the event will be lost. and the joystick can not be opened any more. so I move to preprocess.

File: client/gui/CGuiHandler.cpp [hunk 2]
```diff
@@ -79,6 +79,9 @@ void CGuiHandler::init()
 	renderHandlerInstance = std::make_unique<RenderHandler>();
 	shortcutsHandlerInstance = std::make_unique<ShortcutHandler>();
 	framerateManagerInstance = std::make_unique<FramerateManager>(settings["video"]["targetfps"].Integer());
+
+    // This must be called after SDL_init(), so put after screenHandlerInstance init.
+    inputHandlerInstance->tryOpenGameController();
```
[the remark this answers is not in the document]
Comment (kdmcser) [reply client/gui/CGuiHandler.cpp:84]: yes, put in inputHandlerInstance constructor is much better

File: client/eventsSDL/GameControllerShortcuts.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#pragma once
+
+#include <map>
+#include <SDL.h>
```
[the remark this answers is not in the document]
Comment (kdmcser) [reply client/eventsSDL/GameControllerShortcuts.h:14]: 👌

File: client/eventsSDL/InputSourceGameController.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#pragma once
+
+#include <memory>
```
[the remark this answers is not in the document]
Comment (kdmcser) [reply client/eventsSDL/InputSourceGameController.h:13]: 👌

File: client/eventsSDL/InputSourceGameController.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+#include <SDL.h>
+
+#include "../../lib/Point.h"
```
[the remark this answers is not in the document]
Comment (kdmcser) [reply client/eventsSDL/InputSourceGameController.h:16]: 👌

File: client/eventsSDL/InputHandler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (IvanSavenko) [reply client/eventsSDL/InputHandler.cpp:223]: Ah, in that case - can be kept as it.

**Comment by kdmcser** [comment]
I will define the config file in `config/shortcutsConfig.json` like that:
`
	"joystick": {
		"left-axis": "cursor_motion",
		"right-axis": "cursor_motion",
		"A": ["shortcut1", "shortcut2"],
		"B": [],
		"X": [],
		"Y": [],
		"LB": [],
		"RB": [],
		"LT": [],
		"RT": [],
		"back": [],
		"start": [],
		"up": [],
		"down": [],
		"left": [],
		"right": [],
		"L3" : [],
		"R3": []
	}`

**Comment by IvanSavenko** [comment]
I think it is fine to make sticks hardcoded, so probably no need for this part: `"left-stick": "cursor_motion", "right-stick": "cursor_motion"`.

You also may want to use `SDL_GameControllerGetStringForButton` to get button names instead of using predefined list - some controllers may have additional buttons which hardcoded list might not cover.

**Comment by kdmcser** [comment]
Maybe map roll feature will be implemented , that could be set to left or right stick，so I left the configuration for reserve.
using SDL_GameControllerGetStringForButton seems better, thanks for your advise

**Comment by IvanSavenko** [comment]
Better to reuse existing gesturing system:
- when player starts to move stick, call `GH.events().dispatchGesturePanningStarted(position);`
- after that, continuously call `GH.events().dispatchGesturePanning(startPosition, currentPosition, distance);`
- when player releases the stick, call `GH.events().dispatchGesturePanningEnded(startPosition, finalPosition);`

With this approach not only map scrolling would work, it will also work for any other "scrollable" elements, for example - scenario list.

**Comment by kdmcser** [comment]
👌

**Comment by Laserlicht** [comment]
Tested with Xbox Controller. Worked fine. Good work!

Maybe Target Beta for 1.5?

**Comment by IvanSavenko** [comment]
Hmm... I'll probably merge this into a separate branch, do a quick cleanup / test, and then merge into beta.

**Pull request merged**